hazelcast/src/test/java/com/hazelcast/spi/impl/AbstractInvocationFuture_AbstractTest.java

```diff
@@ -0,0 +1,72 @@
+package com.hazelcast.spi.impl;
+
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.Logger;
+import com.hazelcast.test.HazelcastTestSupport;
+import org.junit.Before;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public abstract class AbstractInvocationFuture_AbstractTest extends HazelcastTestSupport {
+
+    protected ILogger logger;
+    protected Executor executor;
+    protected TestFuture future;
+    protected Object value = "somevalue";
+
+    @Before
+    public void setup() {
+        logger = Logger.getLogger(getClass());
+        executor = Executors.newSingleThreadExecutor();
+        future = new TestFuture();
+    }
+
+
+    class TestFuture extends AbstractInvocationFuture {
+        volatile boolean interruptDetected;
+
+        public TestFuture() {
+            super(AbstractInvocationFuture_AbstractTest.this.executor, AbstractInvocationFuture_AbstractTest.this.logger);
+        }
+
+        public TestFuture(Executor executor, ILogger logger) {
+            super(executor, logger);
+        }
+
+        @Override
+        protected void onInterruptDetected() {
+            interruptDetected = true;
+            complete(new InterruptedException());
```

Comment:
Currently the interrupt is not lost because the interrupt flag is always restored on the thread, no matter the exception being thrown. When the InterruptedException is thrown, the flag is also set. If this is a good or a bad thing is a topic in its own; because due to our exception wrapping on join, the actual InterruptedException gets effectively hidden.